frame/assets/src/lib.rs

```diff
@@ -247,19 +256,594 @@ decl_error! {
 		BalanceLow,
 		/// Balance should be non-zero
 		BalanceZero,
+		/// The signing account has no permission to do the operation.
+		NoPermission,
+		/// The given asset ID is unknown.
+		Unknown,
+		/// The origin account is frozen.
+		Frozen,
+		/// The asset ID is already taken.
+		InUse,
+		/// Too many zombie accounts in use.
+		TooManyZombies,
+		/// Attempt to destroy an asset class when non-zombie, reference-bearing accounts exist.
+		RefsLeft,
+		/// Invalid witness data given.
+		BadWitness,
 	}
 }
 
-decl_storage! {
-	trait Store for Module<T: Trait> as Assets {
-		/// The number of units of assets held by any given account.
-		Balances: map hasher(blake2_128_concat) (T::AssetId, T::AccountId) => T::Balance;
-		/// The next asset identifier up for grabs.
-		NextAssetId get(fn next_asset_id): T::AssetId;
-		/// The total unit supply of an asset.
+pub trait WeightInfo {
+	fn create() -> Weight;
+	fn force_create() -> Weight;
+	fn destroy(z: u32, ) -> Weight;
+	fn force_destroy(z: u32, ) -> Weight;
+	fn mint() -> Weight;
+	fn burn() -> Weight;
+	fn transfer() -> Weight;
+	fn force_transfer() -> Weight;
+	fn freeze() -> Weight;
+	fn thaw() -> Weight;
+	fn transfer_ownership() -> Weight;
+	fn set_team() -> Weight;
+	fn set_max_zombies() -> Weight;
+}
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		type Error = Error<T>;
+
+		fn deposit_event() = default;
+
+		/// Issue a new class of fungible assets from a public origin.
+		///
+		/// This new asset class has no assets initially.
+		///
+		/// The origin must be Signed and the sender must have sufficient funds free.
+		///
+		/// Funds of sender are reserved according to the formula:
+		/// `AssetDepositBase + AssetDepositPerZombie * max_zombies`.
+		///
+		/// Parameters:
+		/// - `id`: The identifier of the new asset. This must not be currently in use to identify
+		/// an existing asset.
+		/// - `owner`: The owner of this class of assets. The owner has full superuser permissions
+		/// over this asset, but may later change and configure the permissions using `transfer_ownership`
+		/// and `set_team`.
+		/// - `max_zombies`: The total number of accounts which may hold assets in this class yet
+		/// have no existential deposit.
+		/// - `min_balance`: The minimum balance of this new asset that any single account must
+		/// have. If an account's balance is reduced below this, then it collapses to zero.
+		///
+		/// Emits `Created` event when successful.
+		///
+		/// Weight: `O(1)`
+		#[weight = T::WeightInfo::create()]
+		fn create(origin,
+			#[compact] id: T::AssetId,
```

Comment:
There's no way to avoid this unless you either store a blacklist of old asset IDs (which leads to state-bloat attacks) or you otherwise force asset IDs into some ever-reducing subset (e.g. which removes the possibility of using the asset ID to freely name the asset).

One possible workaround is to store the blacklist but require an indefinitely-retained deposit, enough to pay for the blacklist entry. It'll likely be a pretty hefty fee...

Reply:
We could just force the asset id to always increment (i.e. have them be based on a counter) and not reuse old ids.
If you don't think that `u32` is enough we can make the index `u64` without loosing much.

Reply:
No - the point is that you can use any type not just one that increments. `[u8;4]` would be an obvious one. `Vec<u8>` another.

Reply:
Ah, ok, support for non-integer types as asset ids makes that suggestion difficult.